File: dotnet/DesignPatterns/BehavioralDesignPatterns/VisitorDesignPattern/VisitorDesignPattern.Example1/ElectronicsProduct.cs
namespace VisitorDesignPattern.Example1;

/// <summary>
/// Concrete element: represents an electronics product.
/// Has an additional property (WarrantyYears) specific to this type.
/// </summary>
public class ElectronicsProduct : IProduct
{
    public string Name { get; }
    public decimal Price { get; }
    public int WarrantyYears { get; }

    public ElectronicsProduct(string name, decimal price, int warrantyYears)
    {
        Name = name;
        Price = price;
        WarrantyYears = warrantyYears;
    }

    public void Accept(IVisitor visitor)
    {
        visitor.Visit(this);
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/VisitorDesignPattern/VisitorDesignPattern.Example1/src/ElectronicsProduct.ts
import { IProduct } from './IProduct';
import { IVisitor } from './IVisitor';

export class ElectronicsProduct implements IProduct {
  readonly name: string;
  readonly price: number;
  readonly warrantyYears: number;

  constructor(name: string, price: number, warrantyYears: number) {
    this.name = name;
    this.price = price;
    this.warrantyYears = warrantyYears;
  }

  accept(visitor: IVisitor): void {
    visitor.visitElectronics(this);
  }
}
